code/game/machinery/computer/prisoner.dm

```diff
@@ -24,118 +21,148 @@
+
+	data["chemicalInfo"] = list()
+	for(var/obj/item/implant/chem/C in GLOB.tracked_implants)
+		var/turf/implant_location = get_turf(C)
+		if(implant_location && implant_location.z != z)
+			continue
+		if(!C.implanted)
+			continue
+		var/list/implant_info = list(
+			"name" = C.imp_in.name,
+			"volume" = C.reagents.total_volume,
+			"uid" = C.UID(),
+		)
+		data["chemicalInfo"] += list(implant_info)
+
+	data["trackingInfo"] = list()
+	for(var/obj/item/implant/tracking/T in GLOB.tracked_implants)
+		var/turf/implant_location = get_turf(T)
+		if(implant_location && implant_location.z != z)
+			continue
+		if(!T.implanted)
+			continue
+
+		var/mob/living/carbon/M = T.imp_in
+		var/loc_display = "Unknown"
+		var/health_display = "OK"
+		var/total_loss = (M.maxHealth - M.health)
+		if(M.stat == DEAD)
+			health_display = "DEAD"
+		else if(total_loss)
+			health_display = "HURT ([total_loss])"
+		if(!istype(implant_location.loc, /turf/space))
+			loc_display = "[get_area(implant_location)]"
+
+		var/list/implant_info = list(
+			"id" = T.id,
+			"subject" = M.name,
+			"location" = loc_display,
+			"health" = health_display,
+			"uid" = T.UID()
+		)
+		data["trackingInfo"] += list(implant_info)
+
+	data["modal"] = ui_modal_data(src)
+
+	return data
+
+/obj/machinery/computer/prisoner/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
 	if(..())
-		return 1
-
-	usr.set_machine(src)
-
-	if(href_list["id"])
-		switch(href_list["id"])
-			if("0")
-				var/obj/item/card/id/prisoner/I = usr.get_active_hand()
-				if(istype(I))
-					usr.drop_item()
-					I.loc = src
-					inserted_id = I
-				else
-					to_chat(usr, "<span class='warning'>No valid ID.</span>")
-			if("1")
-				inserted_id.loc = get_step(src,get_turf(usr))
+		return
+
+	if(ui_act_modal(action, params, ui))
+		return
+	if(ui_login_act(action, params))
+		add_fingerprint(usr)
+		return
+
+	var/mob/living/user = ui.user
+
+	add_fingerprint(usr)
+
+	switch(action)
+		if("id_card")
+			if(inserted_id)
+				inserted_id.forceMove(loc)
 				inserted_id = null
-			if("2")
-				inserted_id.mining_points = 0
-			if("3")
-				var/num = round(input(usr, "Choose prisoner's goal:", "Input an Integer", null) as num|null)
-				if(num >= 0)
-					inserted_id.goal = num
-	if(href_list["inject1"])
-		var/obj/item/implant/I = locate(href_list["inject1"])
-		if(I)	I.activate(1)
-
-	else if(href_list["inject5"])
-		var/obj/item/implant/I = locate(href_list["inject5"])
-		if(I)	I.activate(5)
-
-	else if(href_list["inject10"])
-		var/obj/item/implant/I = locate(href_list["inject10"])
-		if(I)	I.activate(10)
-
-	else if(href_list["lock"])
-		if(src.allowed(usr))
-			authenticated = !authenticated
+				return
+			var/obj/item/card/id/prisoner/I = user.get_active_hand()
+			if(istype(I) && user.drop_item())
+				I.forceMove(src)
+				inserted_id = I
+			else
+				to_chat(user, "<span class='warning'>No valid ID.</span>")
+		if("inject")
+			var/obj/item/implant/chem/implant = locateUID(params["uid"])
```

Comment:
this and the modal that uses UID's will need to have a sanity check, I just noticed this and will fix it soon 👍